// chesslib/lib/brands.js
"use strict";
Object.defineProperties(exports, {
  WHITE: {get: function() {
      return WHITE;
    }},
  BLACK: {get: function() {
      return BLACK;
    }},
  KINGSIDE: {get: function() {
      return KINGSIDE;
    }},
  QUEENSIDE: {get: function() {
      return QUEENSIDE;
    }},
  KING: {get: function() {
      return KING;
    }},
  QUEEN: {get: function() {
      return QUEEN;
    }},
  KNIGHT: {get: function() {
      return KNIGHT;
    }},
  BISHOP: {get: function() {
      return BISHOP;
    }},
  ROOK: {get: function() {
      return ROOK;
    }},
  PAWN: {get: function() {
      return PAWN;
    }},
  DARK: {get: function() {
      return DARK;
    }},
  LIGHT: {get: function() {
      return LIGHT;
    }},
  __esModule: {value: true}
});
var WHITE = 'white';
var BLACK = 'black';
var KINGSIDE = 'kingside';
var QUEENSIDE = 'queenside';
var KING = 'king';
var QUEEN = 'queen';
var KNIGHT = 'knight';
var BISHOP = 'bishop';
var ROOK = 'rook';
var PAWN = 'pawn';
var DARK = 'dark';
var LIGHT = 'light';
