// chesslib/lib/index.js
"use strict";
Object.defineProperties(exports, {
  brands: {get: function() {
      return brands;
    }},
  util: {get: function() {
      return util;
    }},
  Algebraic: {get: function() {
      return $__algebraic__.Algebraic;
    }},
  Board: {get: function() {
      return $__board__.Board;
    }},
  Position: {get: function() {
      return $__position__.Position;
    }},
  Point: {get: function() {
      return $__point__.Point;
    }},
  Line: {get: function() {
      return $__line__.Line;
    }},
  Game: {get: function() {
      return $__game__.Game;
    }},
  FEN: {get: function() {
      return $__fen__.FEN;
    }},
  PGN: {get: function() {
      return $__pgn__.PGN;
    }},
  Rider: {get: function() {
      return $__rider__.Rider;
    }},
  Leaper: {get: function() {
      return $__leaper__.Leaper;
    }},
  King: {get: function() {
      return $__standard__.King;
    }},
  Queen: {get: function() {
      return $__standard__.Queen;
    }},
  Rook: {get: function() {
      return $__standard__.Rook;
    }},
  Bishop: {get: function() {
      return $__standard__.Bishop;
    }},
  Knight: {get: function() {
      return $__standard__.Knight;
    }},
  Pawn: {get: function() {
      return $__standard__.Pawn;
    }},
  fischerandom: {get: function() {
      return $__fischerandom__.fischerandom;
    }},
  doubleFischerandom: {get: function() {
      return $__fischerandom__.doubleFischerandom;
    }},
  ninesixty: {get: function() {
      return $__fischerandom__.ninesixty;
    }},
  ChessError: {get: function() {
      return $__error__.ChessError;
    }},
  MobilityError: {get: function() {
      return $__error__.MobilityError;
    }},
  CheckError: {get: function() {
      return $__error__.CheckError;
    }},
  PromotionError: {get: function() {
      return $__error__.PromotionError;
    }},
  ResultError: {get: function() {
      return $__error__.ResultError;
    }},
  AmbiguityError: {get: function() {
      return $__error__.AmbiguityError;
    }},
  __esModule: {value: true}
});
require('traceur/bin/traceur-runtime.js');
var brands = require('./brands');
var util = require('./util');
var $__algebraic__ = require('./algebraic');
var $__board__ = require('./board');
var $__position__ = require('./position');
var $__point__ = require('./point');
var $__line__ = require('./line');
var $__game__ = require('./game');
var $__fen__ = require('./fen');
var $__pgn__ = require('./pgn');
var $__rider__ = require('./rider');
var $__leaper__ = require('./leaper');
var $__standard__ = require('./standard');
var $__fischerandom__ = require('./fischerandom');
var $__error__ = require('./error');
